### PR TITLE
Add new e2e scenarios that reproduce known issues

### PR DESCRIPTION
**Pull request by noomorph**
**Description:**

There were numerous internal bug reports complaining about 17.6.0 visibility logic rework on iOS, so this pull request aims to add a few cases to make sure the next fix version will include fixes both to problems of 17.5.5 and 17.6.0.

| № | Assertion                                                                         |                                                         Screenshot                                                        |
|:-:|-----------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------:|
| 1 | should be able to scroll tab bar strip with a gradient overlay                    |  |
| 2 | should be able to tap on the main ScrollView regardless of the on-screen keyboard |  |
| 3 | should be able to tap on the badged button                                        |  |
| 4 | should not crash after an attempt to tap an element with an out-of-bounds index:  `await element(by.text('Index')).atIndex(100500).tap();` | moved to #2372 |

To be continued.

### CODE REVIEW

File: detox/test/e2e/28.visibility.test.js [hunk 1]
```diff
@@ -13,9 +13,27 @@ describe('Visibility Cases', () => {
...
 
-  it('should be able to tap on the main ScrollView regardless of the on-screen keyboard', async () => {
+  it(':ios: should be able to tap on the main ScrollView in a blind way, on the center, when the on-screen keyboard is enabled', async () => {
     await element(by.id('screenScroll')).tap();
     await element(by.id('inputExample')).tap();
+
+    const { frame: { x, y, width, height } } = await element(by.id('screenScroll')).getAttributes();
+
+    await element(by.id('screenScroll')).tap({
+      x: Math.floor(x + 0.5 * width),
```
Comment (LeoNatan) [inline detox/test/e2e/28.visibility.test.js:23]: The coordinates are in the element's coordinate space, so it doesn't make sense to add x and y.

**Review by LeoNatan — CHANGES_REQUESTED** [review]

File: detox/test/e2e/28.visibility.test.js (hunk 1 above)
Comment (noomorph) [reply detox/test/e2e/28.visibility.test.js]: Fixed.

Comment (noomorph) [reply detox/test/e2e/28.visibility.test.js]: I still have the error:

```
  ● Visibility Cases › :ios: should be able to tap on the main ScrollView in a blind way, on the center, when the on-screen keyboard is enabled

    Test Failed: View “<RCTScrollView: 0x7ff1cd536f60>” is not visible: view does not pass visibility threshold (75%)
    TIP: To print view hierarchy on failed actions/matches, use loglevel verbose and above.

      19 |     const { elementSafeBounds: bounds } = await element(by.id('screenScroll')).getAttributes();
      20 |
    > 21 |     await element(by.id('screenScroll')).tap({
         |                                          ^
      22 |       x: Math.floor(0.5 * bounds.width),
      23 |       y: Math.floor(0.5 * bounds.height),
      24 |     });

      at Object.it.only (test/e2e/28.visibility.test.js:21:42)
```

Comment (LeoNatan) [reply detox/test/e2e/28.visibility.test.js]: 🤷‍♂️ Is it visible?

Comment (noomorph) [reply detox/test/e2e/28.visibility.test.js]: Yes, but the algorithm thinks it is not. Showing the problem on the example of the internal E2E tests.

| Human | Machine |
|:-----:|:-------:|
|DETOX_VISIBILITY_RCTScrollView  0x7fcaf9dd0af0 _SCREEN|DETOX_VISIBILITY_RCTScrollView  0x7fcaf9dd0af0 _TEST|

Comment (LeoNatan) [reply detox/test/e2e/28.visibility.test.js]: Probably not enough percentage of visibility. Other than implementing a visibility policy setting, I don't see what else can be done here.

Comment (noomorph) [reply detox/test/e2e/28.visibility.test.js]: Can we check absolute value of visible area? E.g., if it is about 40pxx40px area (or whatever Apple HIG say) AND 50% of area is visible, then tap can be executed... 🤔

Comment (LeoNatan) [reply detox/test/e2e/28.visibility.test.js]: I don’t understand what that means.

Comment (LeoNatan) [reply detox/test/e2e/28.visibility.test.js]: Also, keyboard has different proportions on different devices.

Comment (noomorph) [reply detox/test/e2e/28.visibility.test.js]: How do you compare against 75% threshold? What is your formula to calculate the total area vs visible area?

Comment (LeoNatan) [reply detox/test/e2e/28.visibility.test.js]: Count visible pixels, divide by total.

Comment (noomorph) [reply detox/test/e2e/28.visibility.test.js]: When you count visible pixels, can you tell what the biggest visible area region is, and what its dimensions?

File: detox/test/e2e/28.visibility.test.js [hunk 2]
```diff
@@ -24,6 +24,16 @@ describe('Visibility Cases', () => {
     });
   });
 
+  it(':ios: should be able to dismiss the keyboard by scrolling down in the ScrollView', async () => {
```
Comment (LeoNatan) [inline detox/test/e2e/28.visibility.test.js:27]: This won't magically work on its own. It requires that the screen add support for this in its code.

**Comment by LeoNatan** [comment]
Now that #2375 has landed, please merge master and try running your tests. Tap to dismiss keyboard should work.

File: detox/test/e2e/28.visibility.test.js (hunk 2 above)
Comment (noomorph) [reply detox/test/e2e/28.visibility.test.js]: Got it. I've removed the test until the better times come. 😄

**Comment by noomorph** [comment]
@LeoNatan , have you had a chance to look at the case number 3?

**Comment by LeoNatan** [comment]
@noomorph That's a bad implementation. The badge should be part the button hierarchy. There is nothing I can do about it.

**Pull request closed without merging**